lib/nostrum/util.ex

```diff
@@ -152,9 +152,14 @@ defmodule Nostrum.Util do
 
   defp get_new_gateway_url do
     case Api.request(:get, Constants.gateway_bot(), "") do
+      # Handle deeper API error.
+      # For example, not including a token can result in a somewhat opaque 401 unauthorized error
+      {:error, %Nostrum.Error.ApiError{response: %{message: message}}} ->
```

Comment:
We should just match on the error and raise it here, no need to raise just the message as that'll give us a runtime error instead of an `ApiError`.

This will also handle the case where we get an error from `httpoison`, in which case it will probably be whatever it needs to be to be raised.

I guess the comment may need some revising as well in this case.

Reply:
```suggestion
      {:error, error}} ->
        raise(error)
      
```

Reply:
Chalk that up to my lack of knowledge in Elixir error handling. It makes sense to raise the error module. However, since we are making an API call in this function, shouldn't we specifically handle the API failure modes as well as a generic one?

Reply:
Maybe I'm not understanding what you're saying, but we're handling both cases of errors here. 

This handles the Nostrum API error by raising the error struct, and it also handles the generic httpoison error by raising the error struct that they return. Is there another error you're talking about?

Also sorry this took so long to get back to you on, I've been quite under the weather the past few days!

Reply:
No problem, I've been a little busy lately anyway.

I'm referring to your suggested change, where we are replacing the `Nostrum.Error.ApiError` type with the less specific `error`. Shouldn't we continue to match on `Nostrum.Error.ApiError`?

Reply:
The `error` term here would pattern match the `Nostrum.Error.ApiError` struct that is returned by us. It's good to be less specific because this will also pattern match on any httpoison error.

But maybe I'm still not understanding what you're trying to say? 🙃

Reply:
Haha! I'm definitely not explaining it well enough.

I realize that `error` will match any Api-thrown error, as well as any error tuple that it matches against. When we talk about the responsibility of the function, it is responsible for calling our domain (the Api module). Errors that can be raised by that domain might be handled explicitly.

So, in addition to the generic error term, I'm wondering if we should also match the specific Api-thrown error term.

We might not have any specific rescue functionality for `Nostrum.Error.ApiError` now, but at least the function does respect the possible return values and exceptions raised from the module it is calling, and it makes it easier to diagnose what is happening.

Reply:
Hmm, I see where you're coming from now. I fully agree with everything you're saying, but I'd like to make a counter argument.

I think your argument is that the errors that we handle here may have certain outcomes that we wish to handle explicitly, such as an invalid token? I'd argue that any errors we get here (such as a token being invalid) are not things we should try to recover from. In the case of connection errors we'd have to add some logic around this area to handle errors like that. This function is only called when trying to start up the bot and most errors we encounter then are fatal.

Let me know your thoughts and if I was correct!

Reply:
> I think your argument is that the errors that we handle here may have certain outcomes that we wish to handle explicitly, such as an invalid token?

Kind of, yes. In languages like PHP, I usually try to avoid generic exception handling and prefer to rely on exceptions thrown from the facade's (or module's) domain explicitly.

> I'd argue that any errors we get here (such as a token being invalid) are not things we should try to recover from.

Recover, no, but it would be nice to be able to indicate to a developer or operator what the issue is and how they can fix or work around it. I think this is fairly aligned with error messaging in a lot of modern languages and tools. I definitely like the idea of "let it crash" that Erlang/Elixir subscribe to, but this is a different class of error or log message in my opinion.

In any case, I'm 100% okay with avoiding making a change like that right now.

Reply:
We could match on the two different types of errors, but what are we going to do then? We can't bubble them up to some return value without a lot of work, and 99% of users probably wouldn't see them anyways. I don't think we could add anything here to the error that wouldn't be shown my raising the error. Just raising the error is really all we can do IMO.

Maybe we could match on the 401 just for the `ApiError` to say their token is invalid?
We could match on just the error struct itself? That's the only way I can think of that would give us some control over the errors here, but even then the only validation would be that it's an error type we expect which I don't think is helpful. 

For what to do now, I think raising just the message is maybe not so great. I don't know how Discord sources their status code messages, and the point of the error struct is to be raised. Also, we're only limiting the amount of information available to the user. I think we should go with the generic error handling for the moment just to resolve the issue, and if we think of a different way to handle it then we can come back in and change it. 

Let me know your thoughts, and sorry once again for taking so long to get back to you!